solrmonitor/SolrEventListener.go

```diff
@@ -0,0 +1,38 @@
+package solrmonitor
+
+/**
+	The client can register the SolrEventListener to listen to the solr cluster state events from the zookeeper.
+
+	On startup, the events are fired in this order:
+		1. SolrLiveNodesChanged fires for all the solr live nodes (node which keeps indexes). It keeps watch on zk node /solr/live_nodes 's children
+		2. SolrQueryNodesChanged fires for all the solr query nodes (node which coordinates the query execution). It keeps watch on zk node /solr/live_query_nodes 's children
+		3. SolrCollectionsChanged fires for all collections that exist. It keeps watch on zk node /solr/collections 's children.
+			3.1 SolrCollectionStateChanged fires for each collection; keeps watch on zk node /solr/collections/collname/state.json
+			3.2 SolrCollectionReplicaStatesChanged fires if Per replica state(PRS) is enable for collection.
+				If PRS enable then it keeps watch on zk node /solr/collections/collname/state.json 's children. Those children represents state of each replica of collection
+				(reference https://issues.apache.org/jira/browse/SOLR-15052, https://docs.google.com/document/d/1xdxpzUNmTZbk0vTMZqfen9R3ArdHokLITdiISBxCFUg/edit )
+
+	After initialization it delivers events for following condition
+		1. SolrLiveNodesChanged if solr live nodes changes
+		2. SolrQueryNodesChanged if solr query node changes
+		3. SolrCollectionsChanged if number of collections changes in solr cluster
+		4. SolrCollectionStateChanged if collection's number of shards changes, or replica moves, or replica splits.
```

Comment:
Two questions:
1) On collection creation, do you get both `SolrCollectionsChanged` and `SolrCollectionStateChanged` for the new collection?  In what order?
2) On collection deletion, do you get both `SolrCollectionsChanged` and `SolrCollectionStateChanged(nil)` for the deleted collection?  In what order?

Reply:
1. On collection creation, 1st SolrCollectionsChanged will fire, then SolrCollectionStateChanged will fire.
2. On collection deleteion, I think it will same behavior, but let me test that first

Reply:
2. On collection deletion: three events happens
      2.1 PRS state of replica/shard deleted => solr/collections/coll2/state.json children changed
      2.2 state.json get updated => solr/collections/coll2/state.json updated
      2.3 (above two steps happens for each shard/replica)
      2.4 finally event for coll2 deleted = > solr/collections children changed.

Reply:
```
2021/11/29 06:04:06.534134 DEBUG: fs/distrib/zookeeper/zookeeper.go:239: zk EventNodeChildrenChanged event received for /solr/collections/coll2/state.json
2021/11/29 06:04:06.556964 DEBUG: external/com_github_fullstorydev_gosolr/solrmonitor/solrmonitor.go:219: updateCollectionState: children [core_node22:12:A:L core_node11:17:A:L core_node20:10:A:L core_node3:9:A:L core_node14:13:A:L]
2021/11/29 06:04:06.557171 INFO: fs/services/solrmonitor/impl/solrmonitor.go:269: collection coll2 has PRS update [name:"core_node11" version:17 is_active:true is_leader:true  name:"core_node14" version:13 is_active:true is_leader:true  name:"core_node22" version:12 is_active:true is_leader:true  name:"core_node3" version:9 is_active:true is_leader:true  name:"core_node20" version:10 is_active:true is_leader:true ] 
2021/11/29 06:04:06.582454 DEBUG: external/com_github_fullstorydev_gosolr/solrmonitor/zkwatcherman.go:82: ZkWatcherMan data EventNodeDataChanged: /solr/collections/coll2/state.json
2021/11/29 06:04:06.582575 DEBUG: fs/distrib/zookeeper/zookeeper.go:239: zk EventNodeDataChanged event received for /solr/collections/coll2/state.json
2021/11/29 06:04:06.583796 DEBUG: fs/services/solrmonitor/impl/solrmonitor.go:284: SolrCollection Changed coll2 state: started
2021/11/29 06:04:06.650297 DEBUG: external/com_github_fullstorydev_gosolr/solrmonitor/zkwatcherman.go:85: ZkWatcherMan children EventNodeChildrenChanged: /solr/collections/coll2/state.json
2021/11/29 06:04:06.650379 DEBUG: fs/distrib/zookeeper/zookeeper.go:239: zk EventNodeChildrenChanged event received for /solr/collections/coll2/state.json
2021/11/29 06:04:06.651381 DEBUG: external/com_github_fullstorydev_gosolr/solrmonitor/solrmonitor.go:219: updateCollectionState: children [core_node22:12:A:L core_node20:10:A:L core_node3:9:A:L core_node14:13:A:L]
2021/11/29 06:04:06.652090 INFO: fs/services/solrmonitor/impl/solrmonitor.go:269: collection coll2 has PRS update [name:"core_node20" version:10 is_active:true is_leader:true  name:"core_node14" version:13 is_active:true is_leader:true  name:"core_node22" version:12 is_active:true is_leader:true  name:"core_node3" version:9 is_active:true is_leader:true ] 
2021/11/29 06:04:06.674166 DEBUG: external/com_github_fullstorydev_gosolr/solrmonitor/zkwatcherman.go:82: ZkWatcherMan data EventNodeDataChanged: /solr/collections/coll2/state.json
2021/11/29 06:04:06.674485 DEBUG: fs/distrib/zookeeper/zookeeper.go:239: zk EventNodeDataChanged event received for /solr/collections/coll2/state.json
2021/11/29 06:04:06.674512 DEBUG: fs/services/solrmonitor/impl/solrmonitor.go:284: SolrCollection Changed coll2 state: started
2021/11/29 06:04:06.718644 DEBUG: external/com_github_fullstorydev_gosolr/solrmonitor/zkwatcherman.go:85: ZkWatcherMan children EventNodeChildrenChanged: /solr/collections/coll2/state.json
2021/11/29 06:04:06.718853 DEBUG: fs/distrib/zookeeper/zookeeper.go:239: zk EventNodeChildrenChanged event received for /solr/collections/coll2/state.json
2021/11/29 06:04:06.719474 DEBUG: external/com_github_fullstorydev_gosolr/solrmonitor/solrmonitor.go:219: updateCollectionState: children [core_node20:10:A:L core_node3:9:A:L core_node14:13:A:L]
2021/11/29 06:04:06.719619 INFO: fs/services/solrmonitor/impl/solrmonitor.go:269: collection coll2 has PRS update [name:"core_node14" version:13 is_active:true is_leader:true  name:"core_node3" version:9 is_active:true is_leader:true  name:"core_node20" version:10 is_active:true is_leader:true ] 
2021/11/29 06:04:06.743547 DEBUG: external/com_github_fullstorydev_gosolr/solrmonitor/zkwatcherman.go:82: ZkWatcherMan data EventNodeDataChanged: /solr/collections/coll2/state.json
2021/11/29 06:04:06.743658 DEBUG: fs/distrib/zookeeper/zookeeper.go:239: zk EventNodeDataChanged event received for /solr/collections/coll2/state.json
2021/11/29 06:04:06.744351 DEBUG: fs/services/solrmonitor/impl/solrmonitor.go:284: SolrCollection Changed coll2 state: started
2021/11/29 06:04:06.791760 DEBUG: external/com_github_fullstorydev_gosolr/solrmonitor/zkwatcherman.go:85: ZkWatcherMan children EventNodeChildrenChanged: /solr/collections/coll2/state.json
2021/11/29 06:04:06.791947 DEBUG: fs/distrib/zookeeper/zookeeper.go:239: zk EventNodeChildrenChanged event received for /solr/collections/coll2/state.json
2021/11/29 06:04:06.792556 DEBUG: external/com_github_fullstorydev_gosolr/solrmonitor/solrmonitor.go:219: updateCollectionState: children [core_node3:9:A:L core_node14:13:A:L]
2021/11/29 06:04:06.792852 INFO: fs/services/solrmonitor/impl/solrmonitor.go:269: collection coll2 has PRS update [name:"core_node14" version:13 is_active:true is_leader:true  name:"core_node3" version:9 is_active:true is_leader:true ] 
2021/11/29 06:04:06.816423 DEBUG: external/com_github_fullstorydev_gosolr/solrmonitor/zkwatcherman.go:82: ZkWatcherMan data EventNodeDataChanged: /solr/collections/coll2/state.json
2021/11/29 06:04:06.816505 DEBUG: fs/distrib/zookeeper/zookeeper.go:239: zk EventNodeDataChanged event received for /solr/collections/coll2/state.json
2021/11/29 06:04:06.817566 DEBUG: fs/services/solrmonitor/impl/solrmonitor.go:284: SolrCollection Changed coll2 state: started
2021/11/29 06:04:06.862546 DEBUG: external/com_github_fullstorydev_gosolr/solrmonitor/zkwatcherman.go:85: ZkWatcherMan children EventNodeChildrenChanged: /solr/collections/coll2/state.json
2021/11/29 06:04:06.862653 DEBUG: fs/distrib/zookeeper/zookeeper.go:239: zk EventNodeChildrenChanged event received for /solr/collections/coll2/state.json
2021/11/29 06:04:06.863221 DEBUG: external/com_github_fullstorydev_gosolr/solrmonitor/solrmonitor.go:219: updateCollectionState: children [core_node3:9:A:L]
2021/11/29 06:04:06.863574 INFO: fs/services/solrmonitor/impl/solrmonitor.go:269: collection coll2 has PRS update [name:"core_node3" version:9 is_active:true is_leader:true ] 
2021/11/29 06:04:06.886704 DEBUG: external/com_github_fullstorydev_gosolr/solrmonitor/zkwatcherman.go:82: ZkWatcherMan data EventNodeDataChanged: /solr/collections/coll2/state.json
2021/11/29 06:04:06.887331 DEBUG: fs/distrib/zookeeper/zookeeper.go:239: zk EventNodeDataChanged event received for /solr/collections/coll2/state.json
2021/11/29 06:04:06.887483 DEBUG: fs/services/solrmonitor/impl/solrmonitor.go:284: SolrCollection Changed coll2 state: started
2021/11/29 06:04:06.932755 DEBUG: external/com_github_fullstorydev_gosolr/solrmonitor/zkwatcherman.go:85: ZkWatcherMan children EventNodeChildrenChanged: /solr/collections/coll2/state.json
2021/11/29 06:04:06.932964 DEBUG: fs/distrib/zookeeper/zookeeper.go:239: zk EventNodeChildrenChanged event received for /solr/collections/coll2/state.json
2021/11/29 06:04:06.933527 DEBUG: external/com_github_fullstorydev_gosolr/solrmonitor/solrmonitor.go:219: updateCollectionState: children []
2021/11/29 06:04:06.964551 DEBUG: external/com_github_fullstorydev_gosolr/solrmonitor/zkwatcherman.go:82: ZkWatcherMan data EventNodeDataChanged: /solr/collections/coll2/state.json
2021/11/29 06:04:06.964952 DEBUG: fs/distrib/zookeeper/zookeeper.go:239: zk EventNodeDataChanged event received for /solr/collections/coll2/state.json
2021/11/29 06:04:06.966480 DEBUG: fs/services/solrmonitor/impl/solrmonitor.go:284: SolrCollection Changed coll2 state: started
2021/11/29 06:04:07.112889 DEBUG: external/com_github_fullstorydev_gosolr/solrmonitor/zkwatcherman.go:78: ZkWatcherMan EventNodeDeleted: /solr/collections/coll2/state.json
2021/11/29 06:04:07.113066 DEBUG: fs/distrib/zookeeper/zookeeper.go:239: zk EventNodeDeleted event received for /solr/collections/coll2/state.json
2021/11/29 06:04:07.136094 DEBUG: external/com_github_fullstorydev_gosolr/solrmonitor/solrmonitor.go:532: coll2: no data
2021/11/29 06:04:07.136296 DEBUG: fs/services/solrmonitor/impl/solrmonitor.go:284: SolrCollection Changed coll2 state: started
2021/11/29 06:04:07.999124 DEBUG: external/com_github_fullstorydev_gosolr/solrmonitor/zkwatcherman.go:85: ZkWatcherMan children EventNodeChildrenChanged: /solr/collections
2021/11/29 06:04:07.999281 DEBUG: fs/distrib/zookeeper/zookeeper.go:239: zk EventNodeChildrenChanged event received for /solr/collections
2021/11/29 06:04:07.999762 DEBUG: fs/services/solrmonitor/impl/solrmonitor.go:201: solrmonitor (8) -> (7) collections; added=[], removed=[coll2]
```